Content.Server/Medical/Disease/Systems/DiseaseResidueSystem.cs

```diff
@@ -0,0 +1,200 @@
+using System;
+using System.Linq;
+using Content.Shared.Interaction;
+using Content.Shared.Interaction.Events;
+using Content.Shared.Medical.Disease;
+using Content.Shared.Inventory;
+using Content.Shared.Hands.Components;
+using Robust.Shared.Map;
+using Robust.Shared.Collections;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Timing;
+using Content.Shared.StepTrigger.Systems;
+using Content.Shared.Weapons.Melee;
+using Content.Shared.Weapons.Melee.Events;
+
+namespace Content.Server.Medical.Disease.Systems;
+
+/// <summary>
+/// Decays disease residue on tiles/items and infects entities on direct contact.
+/// </summary>
+public sealed class DiseaseResidueSystem : EntitySystem
+{
+    [Dependency] private readonly DiseaseSystem _disease = default!;
+    [Dependency] private readonly IPrototypeManager _prototypes = default!;
+    [Dependency] private readonly EntityLookupSystem _lookup = default!;
+    [Dependency] private readonly SharedInteractionSystem _interaction = default!;
+    [Dependency] private readonly SharedTransformSystem _xform = default!;
+    [Dependency] private readonly InventorySystem _inventory = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<DiseaseCarrierComponent, ContactInteractionEvent>(OnCarrierContact);
+        SubscribeLocalEvent<DiseaseResidueComponent, ContactInteractionEvent>(OnResidueContact);
+        SubscribeLocalEvent<MeleeWeaponComponent, MeleeHitEvent>(OnMeleeHit);
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<DiseaseResidueComponent>();
+        while (query.MoveNext(out var uid, out var residue))
+        {
+            // Decay per-disease intensities.
+            var decay = residue.DecayPerTick * (float)frameTime;
+            var toRemoveAfterDecay = new ValueList<string>();
+            foreach (var kv in residue.Diseases.ToArray())
```

Comment:
This toarray is pointless and also just re-use the valuelist outside of the hot loop.